File: backend/run-test.js
const { spawn } = require('child_process');
const sqlite3 = require('sqlite3').verbose();
const http = require('http');
const path = require('path');

// Step 1: Check database
console.log('\n========================================');
console.log('STEP 1: Checking database for users');
console.log('========================================\n');

const db = new sqlite3.Database(path.resolve(__dirname, 'database.sqlite'), (err) => {
  if (err) {
    console.error('Error opening database:', err.message);
    process.exit(1);
  }
  console.log('✓ Connected to database.sqlite\n');

  db.all('SELECT id, email, firstName, lastName, role, isActive, password FROM users', (err, rows) => {
    if (err) {
      console.error('Error querying users:', err.message);
      db.close();
      process.exit(1);
    }

    console.log(`Found ${rows.length} user(s):\n`);
    rows.forEach(user => {
      console.log(`  ID: ${user.id}`);
      console.log(`  Email: ${user.email}`);
      console.log(`  Name: ${user.firstName} ${user.lastName}`);
      console.log(`  Role: ${user.role}`);
      console.log(`  Active: ${user.isActive}`);
      console.log(`  Password hash: ${user.password.substring(0, 20)}...`);
      console.log('');
    });

    db.close();

    // Step 2: Start server
    console.log('========================================');
    console.log('STEP 2: Starting backend server');
    console.log('========================================\n');

    const server = spawn('node', ['server.js'], {
      cwd: __dirname,
      stdio: 'pipe'
    });

    let serverOutput = '';
    let serverStarted = false;

    server.stdout.on('data', (data) => {
      const output = data.toString();
      serverOutput += output;
      console.log('[SERVER]', output.trim());
      
      if (output.includes('Server running on port')) {
        serverStarted = true;
      }
    });

    server.stderr.on('data', (data) => {
      console.log('[SERVER ERROR]', data.toString().trim());
    });

    // Wait for server to start, then test login
    let waitTime = 0;
    const checkServer = setInterval(() => {
      waitTime += 200;
      
      if (serverStarted || waitTime > 10000) {
        clearInterval(checkServer);
        
        console.log('\n========================================');
        console.log('STEP 3: Testing login endpoint');
        console.log('========================================\n');
        
        testLogin(() => {
          server.kill();
          process.exit(0);
        });
      }
    }, 200);
  });
});

function testLogin(callback) {
  const loginData = JSON.stringify({
    email: 'admin@yaacoub.ma',
    password: 'Admin123!'
  });

  const options = {
    hostname: 'localhost',
    port: 5002,
    path: '/api/auth/login',
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(loginData)
    }
  };

  console.log('Sending login request:');
  console.log(`  URL: http://${options.hostname}:${options.port}${options.path}`);
  console.log(`  Method: ${options.method}`);
  console.log(`  Body: ${loginData}\n`);

  const req = http.request(options, (res) => {
    console.log(`✓ Response received:`);
    console.log(`  Status Code: ${res.statusCode}`);
    console.log(`  Status Text: ${res.statusMessage}`);
    console.log(`  Headers:`);
    Object.entries(res.headers).forEach(([key, value]) => {
      console.log(`    ${key}: ${value}`);
    });

    let data = '';
    res.on('data', (chunk) => {
      data += chunk;
    });

    res.on('end', () => {
      console.log(`\nResponse Body:\n${data}\n`);

      try {
        const parsed = JSON.parse(data);
        console.log('✓ Parsed JSON response:');
        console.log(JSON.stringify(parsed, null, 2));
      } catch (e) {
        console.log('✗ Could not parse JSON response');
      }

      console.log('\n========================================');
      console.log('TEST COMPLETE');
      console.log('========================================\n');

      callback();
    });
  });

  req.on('error', (error) => {
    console.error('✗ Request error:', error.message);
    console.error('Make sure the server is running on port 5002');
    callback();
  });

  req.write(loginData);
  req.end();
}

File: backend/run-test.py
#!/usr/bin/env python3
import sqlite3
import subprocess
import time
import urllib.request
import urllib.error
import json
import sys
import os

# Change to backend directory
os.chdir(r'C:\Users\pc gold\Desktop\wathiqati\backend')

# Step 1: Check database
print('\n' + '='*50)
print('STEP 1: Checking database for users')
print('='*50 + '\n')

try:
    conn = sqlite3.connect('database.sqlite')
    cursor = conn.cursor()
    
    cursor.execute('SELECT id, email, firstName, lastName, role, isActive FROM users')
    users = cursor.fetchall()
    
    print(f'Found {len(users)} user(s):\n')
    for user in users:
        print(f'  ID: {user[0]}')
        print(f'  Email: {user[1]}')
        print(f'  Name: {user[2]} {user[3]}')
        print(f'  Role: {user[4]}')
        print(f'  Active: {user[5]}')
        print()
    
    conn.close()
except Exception as e:
    print(f'Error querying database: {e}')
    sys.exit(1)

# Step 2: Start server
print('='*50)
print('STEP 2: Starting backend server')
print('='*50 + '\n')

server_proc = subprocess.Popen(
    ['node', 'server.js'],
    stdout=subprocess.PIPE,
    stderr=subprocess.STDOUT,
    universal_newlines=True,
    bufsize=1
)

# Capture server output
server_started = False
for _ in range(50):  # Try for up to 10 seconds
    try:
        line = server_proc.stdout.readline()
        if line:
            print(f'[SERVER] {line.strip()}')
            if 'Server running on port' in line:
                server_started = True
                break
    except:
        pass
    time.sleep(0.2)

if not server_started:
    print('Warning: Server may not have started yet, attempting login anyway...')

time.sleep(1)  # Give server a moment to fully initialize

# Step 3: Test login
print('\n' + '='*50)
print('STEP 3: Testing login endpoint')
print('='*50 + '\n')

login_data = json.dumps({
    'email': 'admin@yaacoub.ma',
    'password': 'Admin123!'
}).encode('utf-8')

url = 'http://localhost:5002/api/auth/login'

print(f'Sending login request:')
print(f'  URL: {url}')
print(f'  Method: POST')
print(f'  Body: {login_data.decode()}\n')

try:
    req = urllib.request.Request(
        url,
        data=login_data,
        headers={
            'Content-Type': 'application/json'
        },
        method='POST'
    )
    
    with urllib.request.urlopen(req) as response:
        print(f'✓ Response received:')
        print(f'  Status Code: {response.status}')
        print(f'  Headers:')
        for header, value in response.headers.items():
            print(f'    {header}: {value}')
        
        response_body = response.read().decode('utf-8')
        print(f'\n  Body:\n{response_body}\n')
        
        try:
            parsed = json.loads(response_body)
            print('✓ Parsed JSON response:')
            print(json.dumps(parsed, indent=2))
        except:
            print('✗ Could not parse JSON')

except urllib.error.HTTPError as e:
    print(f'✓ HTTP Error Response received:')
    print(f'  Status Code: {e.code}')
    print(f'  Headers:')
    for header, value in e.headers.items():
        print(f'    {header}: {value}')
    
    error_body = e.read().decode('utf-8')
    print(f'\n  Body:\n{error_body}\n')
    
    try:
        parsed = json.loads(error_body)
        print('✓ Parsed JSON response:')
        print(json.dumps(parsed, indent=2))
    except:
        print('✗ Could not parse JSON')

except urllib.error.URLError as e:
    print(f'✗ Request error: {e.reason}')
    print('Make sure the server is running on port 5002')

except Exception as e:
    print(f'✗ Unexpected error: {e}')

# Read remaining server output
print('\n' + '='*50)
print('STEP 4: Server logs during login attempt')
print('='*50 + '\n')

try:
    while True:
        line = server_proc.stdout.readline()
        if line:
            print(f'[SERVER] {line.rstrip()}')
        else:
            time.sleep(0.1)
except KeyboardInterrupt:
    pass

# Cleanup
server_proc.terminate()
server_proc.wait(timeout=5)

print('\n' + '='*50)
print('TEST COMPLETE')
print('='*50 + '\n')
